pkg/scheduler/objects/queue.go

```diff
@@ -1292,3 +1303,42 @@ func (sq *Queue) String() string {
 	return fmt.Sprintf("{QueuePath: %s, State: %s, StateTime: %x, MaxResource: %s}",
 		sq.QueuePath, sq.stateMachine.Current(), sq.stateTime, sq.maxResource)
 }
+
+func (sq *Queue) incRunningApps() {
+	sq.runningApps++
+	if sq.parent != nil {
+		sq.parent.incRunningApps()
+	}
+}
+
+func (sq *Queue) decRunningApps() {
+	sq.runningApps--
+	if sq.parent != nil {
+		sq.parent.decRunningApps()
+	}
+}
+
+func (sq *Queue) canRun() bool {
+	sq.RLock()
+	defer sq.RUnlock()
```

Comment:
This will lock the current queue and then walk up the tree, locking all parents.
Please follow the same structure as we have for all over: first walk up the tree, check the top and then when unwinding check the current queue. Locking only when the current queue is checked (see getHeadRoom as an example)